spot_wrapper/calibration/calibration_util.py

```diff
@@ -0,0 +1,916 @@
+    images = np.transpose(images, (1, 0))
+
+    return images
+
+
+def save_calibration_parameters(
+    data: Dict,
+    output_path: str,
+    num_images: int,
+    tag: str,
+    parser_args: Optional[argparse.Namespace] = None,
+    unsafe: bool = False,
+) -> None:
+    """
+    Dump the results of a calibration, and the metadata associated with the command that
+    created it, to a file
+
+    Args:
+        data (Dict): The results of the calibration
+        output_path (str): The path/name of what to create
+        num_images (int): How many images were used for this calibration
+        tag (str): What tag to give as the heading/name for this calibration
+        parser_args (Optional[argparse.Namespace], optional): The args that were
+            used to create the calibration. Defaults to None.
+        unsafe (bool, optional): Whether to overwrite existing calibrations of the same name,
+            and to ignore recommended naming scheme. Defaults to False.
+    """
+
+    def flatten_matrix(matrix: np.ndarray) -> List:
+        return matrix.flatten().tolist()
+
+    def process_data_with_nested_dictionaries(
+        data: Dict,
+    ) -> Tuple[Dict, Dict]:
+        cameras = {}
+        relations = {}
+
+        for key, value in data.items():
+            origin_cam, reference_cam = int(key.split("_")[0]), int(key.split("_")[1])
+
+            # Process origin camera data
+            if origin_cam not in cameras:
+                cameras[origin_cam] = {
+                    "camera_matrix": flatten_matrix(value["camera_matrix_origin"]),
+                    "dist_coeffs": flatten_matrix(value["dist_coeffs_origin"]),
+                    "image_dim": flatten_matrix(value["image_dim_origin"]),
+                }
+
+            # Process reference camera data
+            if reference_cam not in cameras:
+                cameras[reference_cam] = {
+                    "camera_matrix": flatten_matrix(value["camera_matrix_reference"]),
+                    "dist_coeffs": flatten_matrix(value["dist_coeffs_reference"]),
+                    "image_dim": flatten_matrix(value["image_dim_reference"]),
+                }
+
+            # Store the rotation and translation in a nested dictionary structure
+            if origin_cam not in relations:
+                relations[origin_cam] = {}
+            relations[origin_cam][reference_cam] = {
+                "R": flatten_matrix(value["R"]),
+                "T": flatten_matrix(value["T"]),
+            }
+
+        return cameras, relations
+
+    # Handle empty or missing tag
+    if not tag:
+        tag = "default"
+
+    # Load existing YAML file if it exists
+    output_path = os.path.expanduser(output_path)
+    if os.path.exists(output_path):
+        with open(output_path, "r") as file:
+            existing_data = yaml.safe_load(file) or {}
+    else:
+        existing_data = {}
+
+    # Check for overwriting existing tag
+    if not unsafe:
```

Comment:
I love this attention to detail :) sometimes one of the most annoying things is when calibration file dont match or something is wrong with the format and it throws an exception.